activerecord/lib/active_record/connection_adapters/sql_type_metadata.rb

```diff
@@ -16,19 +16,22 @@ def initialize(sql_type: nil, type: nil, limit: nil, precision: nil, scale: nil)
 
       def ==(other)
         other.is_a?(SqlTypeMetadata) &&
-          attributes_for_hash == other.attributes_for_hash
+          sql_type == other.sql_type &&
+          type == other.type &&
+          limit == other.limit &&
+          precision == other.precision &&
+          scale == other.scale
       end
       alias eql? ==
 
       def hash
-        attributes_for_hash.hash
+        SqlTypeMetadata.hash ^
+          sql_type.hash ^
+          type.hash ^
+          limit.hash ^
+          precision.hash ^
+          scale.hash
```

Comment:
I wonder if hash collisions will be an issue due to using xor here.

ex. `precision=4 scale=4` will hash the same as `precision=5 scale=5` (because `4 ^ 4 == 5 ^ 5 == 0`)

The same values in a different order would also hash the same

Reply:
Yeah, TBH honest that's a relatively naive implementation.

I looked at what `Array#hash` was doing, and AFAICT it's doing a simple sum of all the members hash codes:

`Array#hash` is here: https://github.com/ruby/ruby/blob/6e0e93e60ec55ab4a28d29d4ad70a3ad7e65c73e/array.c#L4264-L4279

And after unless I'm wrong that `rb_hash_uint` is an alias for `st_hash_uint` which is very straightforward: https://github.com/ruby/ruby/blob/d48783bb0236db505fe1205d1d9822309de53a36/st.c#L1983

Before writing this I looked at some literature or proper hash code implementation (mostly Java stuff), and I saw people suggesting to multiply the individual part with different prime numbers.

However since ruby will happily go over MAX_INT, it might be a bit trickier to do in Ruby.

I'm very much open to suggestions on this part.

Reply:
@jhawthorn I pushed https://github.com/rails/rails/pull/35875/commits/b1b5687d83cdb39ae67619ad6ceb17f465826012 
I'd love your opinion on wether the tradeoff is worth it or not.

Also I'm starting to wonder if collisions are that much of a big deal. Out of 7k columns, we only have 91 distinct `SqlTypeMetadata`. Not sure if these number would hold in other apps, but if so, collisions impact would be negligible.

Reply:
We could divide (or `>>` shift right) to avoid going having to allocate an Integer. I don't think the prime number is necessary either, since we're dealing with a hash.

Reply:
> We could divide (or `>>` shift right)

That makes so much sense... I'll do that.